src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests/CustomConverterTests.HandleNull.cs

```diff
@@ -0,0 +1,484 @@
+
+            val = JsonSerializer.Deserialize<Uri>("null", options);
+            Assert.Null(val);
+            Assert.Equal("null", JsonSerializer.Serialize(val, options));
+        }
+
+        private class UriNullConverter_SpecialCaseNull : JsonConverter<Uri>
+        {
+            public override Uri Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                if (reader.TokenType == JsonTokenType.Null)
+                {
+                    return new Uri("https://default");
+                }
+
+                throw new JsonException();
+            }
+
+            public override void Write(Utf8JsonWriter writer, Uri value, JsonSerializerOptions options)
+            {
+                if (value == null)
+                {
+                    writer.WriteStringValue("https://default");
+                    return;
+                }
+
+                throw new NotSupportedException();
+            }
+        }
+
+        [Fact]
+        public static void ReferenceTypeConverter_OptOut()
+        {
+            // Per null handling opt-out, serializer handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new UriNullConverter_OptOut());
+
+            Uri val = JsonSerializer.Deserialize<Uri>("null", options);
+            Assert.Null(val);
+            Assert.Equal("null", JsonSerializer.Serialize(val, options));
+        }
+
+        private class UriNullConverter_OptOut : UriNullConverter_SpecialCaseNull
+        {
+            public override bool HandleNull => false;
+        }
+
+        [Fact]
+        public static void ReferenceTypeConverter_NullOptIn()
+        {
+            // Per null handling opt-in, converter handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new UriNullConverter_NullOptIn());
+
+            Uri val = JsonSerializer.Deserialize<Uri>("null", options);
+            Assert.Equal(new Uri("https://default"), val);
+
+            val = null;
+            Assert.Equal(@"""https://default""", JsonSerializer.Serialize(val, options));
+        }
+
+        private class UriNullConverter_NullOptIn : UriNullConverter_SpecialCaseNull
+        {
+            public override bool HandleNull => true;
+        }
+
+        [Fact]
+        public static void ComplexReferenceTypeConverter_NoOverride()
+        {
+            // Baseline
+            Point_2D obj = JsonSerializer.Deserialize<Point_2D>("null");
+            Assert.Null(obj);
+            Assert.Equal("null", JsonSerializer.Serialize(obj));
+
+            // Per null handling default value for reference types (false), serializer handles null.
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new PointClassConverter_SpecialCaseNull());
+
+            obj = JsonSerializer.Deserialize<Point_2D>("null", options);
```

Comment:
Doesn't the converter return `Point_2D(-1, -1)` (not null) for TokenType.Null?

Reply:
By default, reference types don't handle `null` tokens, so the serializer handles `null` (sets the default value). The comment a couple lines above mentions this.